common/optimiser/Optimiser.java

```diff
@@ -83,67 +110,76 @@ public boolean isError() {
         constraints.forEach(constraint -> constraint.initialise(solver));
-        setSolverModel();
-        hasSolver = true;
+        setConstraintCoefficients();
+        setObjective();
     }
 
-    private void setSolverModel() {
-        constraints.forEach(OptimiserConstraint::updateCoefficients);
-        setObjective();
-        setVariableValues();
+    private void setConstraintCoefficients() {
+        constraints.forEach(OptimiserConstraint::setCoefficients);
+        isConstraintsUpToDate = false;
     }
 
     private void releaseSolver() {
         constraints.forEach(OptimiserConstraint::release);
         variables.forEach(OptimiserVariable::release);
         parameters.delete();
         solver.delete();
-        hasSolver = false;
+        solver = null;
+    }
+
+    private void setSolutionAsHints() {
+        assert iterate(variables).allMatch(OptimiserVariable::hasValue);
+        MPVariable[] mpVariables = new MPVariable[variables.size()];
+        double[] hints = new double[variables.size()];
+        for (int i = 0; i < variables.size(); i++) {
+            mpVariables[i] = variables.get(i).mpVariable();
+            hints[i] = variables.get(i).valueAsDouble();
+        }
+        solver.setHint(mpVariables, hints);
     }
 
     private void setObjective() {
         objectiveCoefficients.forEach((var, coeff) -> solver.objective().setCoefficient(var.mpVariable(), coeff));
     }
 
     public double objectiveValue() {
+        assert !objectiveCoefficients.isEmpty();
+        if (objectiveValue == null) objectiveValue = evaluateObjective();
         return objectiveValue;
     }
 
-    private void setVariableValues() {
-        assert iterate(variables).allMatch(OptimiserVariable::hasValue);
-        MPVariable[] mpVariables = new MPVariable[variables.size()];
-        double[] initialisations = new double[variables.size()];
-        for (int i = 0; i < variables.size(); i++) {
-            mpVariables[i] = variables.get(i).mpVariable();
-            initialisations[i] = variables.get(i).hint();
+    private double evaluateObjective() {
+        double value = 0.0;
+        for (Map.Entry<OptimiserVariable<?>, Double> term : objectiveCoefficients.entrySet()) {
+            value += term.getKey().valueAsDouble() * term.getValue();
         }
-        solver.setHint(mpVariables, initialisations);
+        return value;
     }
 
     public void setObjectiveCoefficient(OptimiserVariable<?> var, double coeff) {
-        objectiveCoefficients.put(var, coeff);
-        if (hasSolver) solver.objective().setCoefficient(var.mpVariable(), coeff);
-        else if (isOptimal()) status = Status.FEASIBLE;
+        Double prevCoeff = objectiveCoefficients.put(var, coeff);
+        if (!Objects.equals(coeff, prevCoeff)) {
+            if (solver != null) solver.objective().setCoefficient(var.mpVariable(), coeff);
+            else if (isOptimal()) status = Status.FEASIBLE;
+            objectiveValue = null;
```

Comment:
because we likely set many coefficients at a time, we just `null` the objective value and re-compute it when asked to the next time via `objectiveValue()`